# Ensemble Learning

## Initial Imports

In [19]:
import warnings
warnings.filterwarnings('ignore')

In [20]:
import numpy as np
import pandas as pd
from pathlib import Path
from collections import Counter

In [21]:
from sklearn.metrics import balanced_accuracy_score
from sklearn.metrics import confusion_matrix
from imblearn.metrics import classification_report_imbalanced

## Read the CSV and Perform Basic Data Cleaning

In [22]:
# Load the data
file_path = Path('Resources/LoanStats_2019Q1.csv')
df = pd.read_csv(file_path)

# Preview the data
df.head()

loan_amnt  int_rate  installment home_ownership  annual_inc  \
0    10500.0    0.1719       375.35           RENT     66000.0   
1    25000.0    0.2000       929.09       MORTGAGE    105000.0   
2    20000.0    0.2000       529.88       MORTGAGE     56000.0   
3    10000.0    0.1640       353.55           RENT     92000.0   
4    22000.0    0.1474       520.39       MORTGAGE     52000.0   

  verification_status   issue_d loan_status pymnt_plan    dti  ...  \
0     Source Verified  Mar-2019    low_risk          n  27.24  ...   
1            Verified  Mar-2019    low_risk          n  20.23  ...   
2            Verified  Mar-2019    low_risk          n  24.26  ...   
3            Verified  Mar-2019    low_risk          n  31.44  ...   
4        Not Verified  Mar-2019    low_risk          n  18.76  ...   

   pct_tl_nvr_dlq  percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  \
0            85.7             100.0                   0.0        0.0   
1            91.2              50.0                   1.0        0.0   
2            66.7              50.0                   0.0        0.0   
3           100.0              50.0                   1.0        0.0   
4           100.0               0.0                   0.0        0.0   

   tot_hi_cred_lim  total_bal_ex_mort total_bc_limit  \
0          65687.0            38199.0         2000.0   
1         271427.0            60641.0        41200.0   
2          60644.0            45684.0         7500.0   
3          99506.0            68784.0        19700.0   
4         219750.0            25919.0        27600.0   

   total_il_high_credit_limit  hardship_flag  debt_settlement_flag  
0                     61987.0              N                     N  
1                     49197.0              N                     N  
2                     43144.0              N                     N  
3                     76506.0              N                     N  
4                     20000.0              N                     N  

[5 rows x 86 columns]

In [23]:
# One Hot Encoding - left in here but using dummies below
#df = df.apply(LabelEncoder().fit_transform)
#df

## Split the Data into Training and Testing

In [24]:
# Create our features
X = df.drop(columns="loan_status")
X = pd.get_dummies(X)

# Create our target
y = df['loan_status']

In [25]:
X.describe()

loan_amnt      int_rate   installment    annual_inc           dti  \
count  68817.000000  68817.000000  68817.000000  6.881700e+04  68817.000000   
mean   16677.594562      0.127718    480.652863  8.821371e+04     21.778153   
std    10277.348590      0.048130    288.062432  1.155800e+05     20.199244   
min     1000.000000      0.060000     30.890000  4.000000e+01      0.000000   
25%     9000.000000      0.088100    265.730000  5.000000e+04     13.890000   
50%    15000.000000      0.118000    404.560000  7.300000e+04     19.760000   
75%    24000.000000      0.155700    648.100000  1.040000e+05     26.660000   
max    40000.000000      0.308400   1676.230000  8.797500e+06    999.000000   

        delinq_2yrs  inq_last_6mths      open_acc       pub_rec  \
count  68817.000000    68817.000000  68817.000000  68817.000000   
mean       0.217766        0.497697     12.587340      0.126030   
std        0.718367        0.758122      6.022869      0.336797   
min        0.000000        0.000000      2.000000      0.000000   
25%        0.000000        0.000000      8.000000      0.000000   
50%        0.000000        0.000000     11.000000      0.000000   
75%        0.000000        1.000000     16.000000      0.000000   
max       18.000000        5.000000     72.000000      4.000000   

           revol_bal  ...  issue_d_Mar-2019  pymnt_plan_n  \
count   68817.000000  ...      68817.000000       68817.0   
mean    17604.142828  ...          0.177238           1.0   
std     21835.880400  ...          0.381873           0.0   
min         0.000000  ...          0.000000           1.0   
25%      6293.000000  ...          0.000000           1.0   
50%     12068.000000  ...          0.000000           1.0   
75%     21735.000000  ...          0.000000           1.0   
max    587191.000000  ...          1.000000           1.0   

       initial_list_status_f  initial_list_status_w  next_pymnt_d_Apr-2019  \
count           68817.000000           68817.000000           68817.000000   
mean                0.123879               0.876121               0.383161   
std                 0.329446               0.329446               0.486161   
min                 0.000000               0.000000               0.000000   
25%                 0.000000               1.000000               0.000000   
50%                 0.000000               1.000000               0.000000   
75%                 0.000000               1.000000               1.000000   
max                 1.000000               1.000000               1.000000   

       next_pymnt_d_May-2019  application_type_Individual  \
count           68817.000000                 68817.000000   
mean                0.616839                     0.860340   
std                 0.486161                     0.346637   
min                 0.000000                     0.000000   
25%                 0.000000                     1.000000   
50%                 1.000000                     1.000000   
75%                 1.000000                     1.000000   
max                 1.000000                     1.000000   

       application_type_Joint App  hardship_flag_N  debt_settlement_flag_N  
count                68817.000000          68817.0                 68817.0  
mean                     0.139660              1.0                     1.0  
std                      0.346637              0.0                     0.0  
min                      0.000000              1.0                     1.0  
25%                      0.000000              1.0                     1.0  
50%                      0.000000              1.0                     1.0  
75%                      0.000000              1.0                     1.0  
max                      1.000000              1.0                     1.0  

[8 rows x 95 columns]

In [26]:
X

loan_amnt  int_rate  installment  annual_inc    dti  delinq_2yrs  \
0        10500.0    0.1719       375.35     66000.0  27.24          0.0   
1        25000.0    0.2000       929.09    105000.0  20.23          0.0   
2        20000.0    0.2000       529.88     56000.0  24.26          0.0   
3        10000.0    0.1640       353.55     92000.0  31.44          0.0   
4        22000.0    0.1474       520.39     52000.0  18.76          0.0   
...          ...       ...          ...         ...    ...          ...   
68812    10000.0    0.1502       346.76     26000.0   9.60          0.0   
68813    12000.0    0.2727       368.37     63000.0  29.07          0.0   
68814     5000.0    0.1992       185.62     52000.0  14.86          0.0   
68815    40000.0    0.0646      1225.24    520000.0   9.96          0.0   
68816    16000.0    0.1131       350.36     72000.0   7.02          2.0   

       inq_last_6mths  open_acc  pub_rec  revol_bal  ...  issue_d_Mar-2019  \
0                 0.0       8.0      0.0     1609.0  ...                 1   
1                 0.0      17.0      1.0    18368.0  ...                 1   
2                 0.0       8.0      0.0    13247.0  ...                 1   
3                 1.0      10.0      1.0    17996.0  ...                 1   
4                 1.0      14.0      0.0     9091.0  ...                 1   
...               ...       ...      ...        ...  ...               ...   
68812             0.0       9.0      0.0     2684.0  ...                 0   
68813             0.0       8.0      0.0    13314.0  ...                 0   
68814             0.0       5.0      1.0     3715.0  ...                 0   
68815             1.0      21.0      0.0    59529.0  ...                 0   
68816             0.0      12.0      1.0    11882.0  ...                 0   

       pymnt_plan_n  initial_list_status_f  initial_list_status_w  \
0                 1                      0                      1   
1                 1                      0                      1   
2                 1                      0                      1   
3                 1                      0                      1   
4                 1                      0                      1   
...             ...                    ...                    ...   
68812             1                      0                      1   
68813             1                      0                      1   
68814             1                      0                      1   
68815             1                      1                      0   
68816             1                      0                      1   

       next_pymnt_d_Apr-2019  next_pymnt_d_May-2019  \
0                          0                      1   
1                          0                      1   
2                          0                      1   
3                          0                      1   
4                          0                      1   
...                      ...                    ...   
68812                      0                      1   
68813                      0                      1   
68814                      0                      1   
68815                      0                      1   
68816                      0                      1   

       application_type_Individual  application_type_Joint App  \
0                                1                           0   
1                                1                           0   
2                                1                           0   
3                                1                           0   
4                                1                           0   
...                            ...                         ...   
68812                            1                           0   
68813                            1                           0   
68814                            1                           0   
68815              

In [27]:
# Check the balance of our target values
df['loan_status'].value_counts()

low_risk     68470
high_risk      347
Name: loan_status, dtype: int64

In [28]:
# Split the X and y into X_train, X_test, y_train, y_test
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1, stratify=y)
X_train.shape

(51612, 95)

## Data Pre-Processing

Scale the training and testing data using the `StandardScaler` from `sklearn`. Remember that when scaling the data, you only scale the features data (`X_train` and `X_testing`).

In [29]:
# Create the StandardScaler instance
from sklearn.preprocessing import StandardScaler
data_scaler = StandardScaler()
data_scaler

StandardScaler()

In [30]:
# Fit the Standard Scaler with the training data
# When fitting scaling functions, only train on the training dataset
data_scaler.fit(X_train)

StandardScaler()

In [31]:
# Scale the training and testing data
X_train_scaled = data_scaler.transform(X_train)
X_test_scaled = data_scaler.transform(X_test)

In [32]:
X_train_scaled

array([[-0.64978133, -0.30327574, -0.52669276, ..., -0.40497582,
         0.        ,  0.        ],
       [ 0.42378352, -0.42616892,  0.71269346, ..., -0.40497582,
         0.        ,  0.        ],
       [ 1.79013878, -0.82400957,  2.19335392, ..., -0.40497582,
         0.        ,  0.        ],
       ...,
       [ 0.81417074, -0.36368086,  1.17925683, ..., -0.40497582,
         0.        ,  0.        ],
       [-0.25939411, -0.42616892, -0.08220905, ..., -0.40497582,
         0.        ,  0.        ],
       [-0.16179731, -1.196855  , -0.05849257, ..., -0.40497582,
         0.        ,  0.        ]])

## Ensemble Learners

In this section, you will compare two ensemble algorithms to determine which algorithm results in the best performance. You will train a Balanced Random Forest Classifier and an Easy Ensemble classifier . For each algorithm, be sure to complete the following steps:

1. Train the model using the training data. 
2. Calculate the balanced accuracy score from sklearn.metrics.
3. Display the confusion matrix from sklearn.metrics.
4. Generate a classication report using the `imbalanced_classification_report` from imbalanced-learn.
5. For the Balanced Random Forest Classifier only, print the feature importance sorted in descending order (most important feature to least important) along with the feature score

Note: Use a random state of 1 for each algorithm to ensure consistency between tests

### Balanced Random Forest Classifier

In [33]:
# Resample the training data with the BalancedRandomForestClassifier
from sklearn.ensemble import RandomForestClassifier
rf = RandomForestClassifier(n_estimators=100, random_state=1)
rf = rf.fit(X_train, y_train)

In [34]:
# Calculated the balanced accuracy score
predictions = rf.predict(X_test)

acc_score = balanced_accuracy_score(y_test, predictions)
acc_score

0.67209249388625

In [35]:
# Display the confusion matrix
cm = confusion_matrix(y_test, predictions)
cm_df = pd.DataFrame(cm, index=["Actual 0", "Actual 1"], columns=["Predicted 0", "Predicted 1"])
cm_df

Predicted 0  Predicted 1
Actual 0           30           57
Actual 1           11        17107

In [36]:
# Print the imbalanced classification report
print("Classification Report")
print(classification_report_imbalanced(y_test, predictions))

Classification Report
                   pre       rec       spe        f1       geo       iba       sup

  high_risk       0.73      0.34      1.00      0.47      0.59      0.32        87
   low_risk       1.00      1.00      0.34      1.00      0.59      0.37     17118

avg / total       1.00      1.00      0.35      1.00      0.59      0.37     17205



In [37]:
# List the features sorted in descending order by feature importance
importances = rf.feature_importances_

importances_sorted = sorted(zip(rf.feature_importances_, X.columns), reverse=True)
print("Top 20 Important Features")
importances_sorted[:20]

Top 20 Important Features


[(0.08758715176950288, 'total_rec_prncp'),
 (0.07770386368390605, 'total_pymnt'),
 (0.0668123203275906, 'total_pymnt_inv'),
 (0.06459395200819652, 'last_pymnt_amnt'),
 (0.05039235474187962, 'total_rec_int'),
 (0.02136336078215045, 'out_prncp_inv'),
 (0.021208274791883988, 'installment'),
 (0.01897321061620999, 'out_prncp'),
 (0.01626915816044429, 'dti'),
 (0.015749878157203166, 'total_bal_ex_mort'),
 (0.015582867584390167, 'mo_sin_old_rev_tl_op'),
 (0.01453797246616538, 'total_rec_late_fee'),
 (0.014132027435623728, 'avg_cur_bal'),
 (0.014110841747070672, 'max_bal_bc'),
 (0.013983181108134525, 'revol_bal'),
 (0.013859883554018463, 'mo_sin_old_il_acct'),
 (0.013841987813126443, 'bc_util'),
 (0.01380522133571516, 'loan_amnt'),
 (0.013366877033695225, 'bc_open_to_buy'),
 (0.013079969483511075, 'annual_inc')]

### Easy Ensemble Classifier

In [38]:
# Train the Classifier
# Easy ensemble classifier i.e. boosted tree - Gradient Boosted Classifier
from sklearn.ensemble import GradientBoostingClassifier
   
# Choose a learning rate and create classifier
classifier = GradientBoostingClassifier(n_estimators=100,
                                    learning_rate=0.8,
                                    max_features=5,
                                    max_depth=3,
                                    random_state=0)

# Fit the model
classifier.fit(X_train_scaled, y_train.ravel())

# Make Prediction
predictions = classifier.predict(X_test_scaled)

In [39]:
# Calculated the balanced accuracy score
acc_score = balanced_accuracy_score(y_test, predictions)
print(f"Balanced Accuracy Score : {acc_score}")

Balanced Accuracy Score : 0.6757365709013703


In [40]:
# Display the confusion matrix
cm = confusion_matrix(y_test, predictions)
cm_df = pd.DataFrame(cm, index=["Actual 0", "Actual 1"], columns=["Predicted 0", "Predicted 1"])

# Displaying results
display(cm_df)

Predicted 0  Predicted 1
Actual 0           31           56
Actual 1           83        17035

In [41]:
# Print the imbalanced classification report
print("Imbalanced Classification Report")
print(classification_report_imbalanced(y_test, predictions))

Imbalanced Classification Report
                   pre       rec       spe        f1       geo       iba       sup

  high_risk       0.27      0.36      1.00      0.31      0.60      0.33        87
   low_risk       1.00      1.00      0.36      1.00      0.60      0.38     17118

avg / total       0.99      0.99      0.36      0.99      0.60      0.38     17205



### Final Questions

1. Which model had the best balanced accuracy score?

    Gradient Boosted Classifier at 0.6757365709013703

2. Which model had the best recall score?

    Balanced Random Forest Classifier at 1.00

3. Which model had the best geometric mean score?

    Boosted Gradient Classifier at 0.60

4. What are the top three features?

    #1. 0.08758715176950288, 'total_rec_prncp'
    
    #2. 0.07770386368390605, 'total_pymnt'
    
    #3. 0.0668123203275906, 'total_pymnt_inv'